### PR TITLE
Process movements for offenders moving to immigration centres

### PR DESCRIPTION
**Pull request by theKHutDeveloper**
This PR is the result of a bug that was showing incorrect numbers of Prison Offender Manager (POM) allocations in a particular prison. Looking into this issue we were able to identify that a couple of POM's had not been de-allocated when an offender had been released or transferred,
and this was resulting in the incorrect POM allocation counts.

The MovementService handles de-allocating POMs when an offender moves prison, so we ran a check across all prisons to find out if there were more situations where POM's were showing up as allocated to an offender that was no longer at the POM's prison.

We found approximately 140 other cases and the vast majority were due to offenders moving from prison to an immigration centre, awaiting deportation. The MovementService did not process these movements as immigration centres are not part of the prison estate and as a result these offenders were not being de-allocated from their POM when they moved.

This commit intends to fix this issue with the MovementService by de-allocating the POM and deleting the CaseInformation when the offender is being transferred or released to an immigration centre - signified by the agency code "IMM" (Immigration) or "MHI" (Morton Hall Immigration Removal Centre).

### CODE REVIEW

File: spec/services/movement_service_spec.rb [hunk 1]
```diff
@@ -189,4 +189,135 @@
...
+
+    let(:immigration_movement) do
+      create(:movement, offender_no: 'G4273GI', direction_code: direction_code, create_date_time: Date.new(2020, 1, 6),
+                        movement_type: movement_type, from_agency: from_agency, to_agency: to_agency)
+    end
+
+    context 'when movement is a transfer' do
+      let(:case_info) { create(:case_information, nomis_offender_id: 'G4273GI') }
+
+      context 'when the from_agency is MHI' do
```
Comment (kjdchapman) [inline spec/services/movement_service_spec.rb]: I'd write out what MHI means in full here, which I think is 
```ruby
context 'when the movement is to an immigration centre' do
``` 
or similar?

File: app/services/movement_service.rb [hunk 2]
```diff
@@ -88,17 +100,31 @@ def self.process_transfer(transfer)
...
     true
   end
+  # rubocop:enable Metrics/MethodLength
 
   # When an offender is released, we can no longer rely on their
   # case information (in case they come back one day), and we
   # should de-activate any current allocations.
   def self.process_release(release)
-    return false unless release.to_agency == 'OUT' && release.from_prison?
+    return false unless release.to_agency == 'OUT'
```
[the remark this answers is not in the document]
Comment (theKHutDeveloper) [reply app/services/movement_service.rb]: Added this in

File: app/services/movement_service.rb [hunk 3]
```diff
@@ -52,6 +53,17 @@ def self.process_transfer(transfer)
...
 
+    # Remove the case information and deallocate offender
+    # when the movement is from immigration or a detention centre
+    # and is not going back to a prison OR
+    # when the movement is from a prison to an immigration or detention centre
+    if ((%w[IMM MHI].include? transfer.from_agency) && !transfer.to_prison?) ||
+    ((%w[IMM MHI].include? transfer.to_agency) && transfer.from_prison?)
+      release_offender(transfer.offender_no)
+
+      return true
```
[the remark this answers is not in the document]
Comment (theKHutDeveloper) [reply app/services/movement_service.rb]: Maybe a conversation for another day?

File: app/services/movement_service.rb [hunk 4]
```diff
@@ -52,6 +53,17 @@ def self.process_transfer(transfer)
       OpenPrisonTransferJob.perform_later(transfer.to_json)
     end
 
+    # Remove the case information and deallocate offender
```
[the remark this answers is not in the document]
Comment (theKHutDeveloper) [reply app/services/movement_service.rb]: Happy to discuss this further when you return

File: spec/services/movement_service_spec.rb (hunk 1 above)
Comment (theKHutDeveloper) [reply spec/services/movement_service_spec.rb]: I don't think this is necessary as it is explained in the code 🤷‍♀ but again, happy to discuss

**Pull request merged**